File: client/packages/domain/lib/src/rental_request/entities/rental_request_entity.dart
enum RentalRequestStatus { pending, accepted, rejected, contracted }

class MemberInfo {
  const MemberInfo({
    required this.fullName,
    required this.isRoomLeader,
    this.phone,
    this.identityNumber,
    this.identityImageUrl,
    this.email,
    this.address,
    this.dateOfBirth,
  });

  factory MemberInfo.fromJson(Map<String, dynamic> json) => MemberInfo(
        fullName: json['fullName'] as String,
        isRoomLeader: json['isRoomLeader'] as bool? ?? false,
        phone: json['phone'] as String?,
        identityNumber: json['identityNumber'] as String?,
        identityImageUrl: json['identityImageUrl'] as String?,
        email: json['email'] as String?,
        address: json['address'] as String?,
        dateOfBirth: json['dateOfBirth'] as String?,
      );

  final String fullName;
  final bool isRoomLeader;
  final String? phone;
  final String? identityNumber;
  final String? identityImageUrl;
  final String? email;
  final String? address;
  final String? dateOfBirth;

  Map<String, dynamic> toJson() => {
        'fullName': fullName,
        'isRoomLeader': isRoomLeader,
        if (phone != null) 'phone': phone,
        if (identityNumber != null) 'identityNumber': identityNumber,
        if (identityImageUrl != null) 'identityImageUrl': identityImageUrl,
        if (email != null) 'email': email,
        if (address != null) 'address': address,
        if (dateOfBirth != null) 'dateOfBirth': dateOfBirth,
      };
}

class VehicleInfo {
  const VehicleInfo({
    required this.type,
    required this.plate,
    required this.quantity,
  });

  factory VehicleInfo.fromJson(Map<String, dynamic> json) => VehicleInfo(
        type: json['type'] as String,
        plate: json['plate'] as String,
        quantity: json['quantity'] as int? ?? 1,
      );

  final String type;
  final String plate;
  final int quantity;

  Map<String, dynamic> toJson() => {
        'type': type,
        'plate': plate,
        'quantity': quantity,
      };
}

abstract class RentalRequestEntity {
  RentalRequestEntity({
    required this.id,
    required this.tenantId,
    required this.roomId,
    required this.status,
    required this.createdAt,
    required this.updatedAt,
    this.note,
    this.memberInfo = const [],
    this.parkingInfo = const [],
  });

  final String id;
  final String tenantId;
  final String roomId;
  final String? note;
  final List<MemberInfo> memberInfo;
  final List<VehicleInfo> parkingInfo;
  final RentalRequestStatus status;
  final String createdAt;
  final String updatedAt;
}

File: server/src/modules/rental-requests/domain/entities/rental-request.entity.ts
export type RentalRequestStatus =
  | 'pending'
  | 'accepted'
  | 'rejected'
  | 'contracted';

export interface MemberInfo {
  fullName: string;
  phone?: string;
  identityNumber?: string;
  identityImageUrl?: string;
  email?: string;
  address?: string;
  dateOfBirth?: string;
  isRoomLeader: boolean;
}

export interface VehicleInfo {
  type: string;
  plate: string;
  quantity: number;
}

export class RentalRequestEntity {
  constructor(
    public readonly id: string,
    public readonly tenantId: string,
    public readonly roomId: string,
    public readonly landlordId: string | null,
    public readonly note: string | null,
    public readonly memberInfo: MemberInfo[],
    public readonly parkingInfo: VehicleInfo[],
    public readonly status: RentalRequestStatus,
    public readonly createdAt: Date,
    public readonly updatedAt: Date,
    public readonly roomTitle: string | null = null,
  ) {}
}
